src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs

```diff
@@ -29,15 +27,15 @@ namespace Microsoft.CodeAnalysis.LanguageServer.Handler
     /// complex data, such as edits and commands, to be computed only when necessary
     /// (i.e. when hovering/previewing a code action).
     /// </summary>
-    [ExportLspMethod(MSLSPMethods.TextDocumentCodeActionResolveName, mutatesSolutionState: false), Shared]
+    [LspMethod(MSLSPMethods.TextDocumentCodeActionResolveName, mutatesSolutionState: false)]
```

Comment:
I split the metadata into two parts.
1.  The language name - this is attached to the ExportProvider attribute.  This way we can lazily import the providers per language to avoid realizing all the dependencies if for example XAML isn't loaded.
2.  The LSP method metadata - this is attached to each individual handler as each handler is responsible for a single LSP method.

Note that this is not an export, it's just defining metadata.  I could define it on the provider, but then I need to deal with multiple metadata attributes per provider and it gets more complex.  This keeps everything nicely grouped.

Reply:
The problem with having a single handler do multiple requests is:
1.  For example, code actions share the cache between all three requests.  But only one of them is mutating.  A single class can therefore would need to either support multiple annotations (and therefore cannot be part of an export annotation) or the mutating metadata could not be part of the annotation.
2.  a single type cannot support multiple requests with the same input/output types (granted we don't have any right now, but it is a possibility).

I'm not quite sure I get what you want in the dictionary and the advantage though - would that be of LSP method -> handler instance?

Reply:
For the dictionary we would still need to store the metadata then to pick which one is mutating,maybe store metadata -> handler?

Reply:
Yeah, I was thinking method -> handler. Essentially I see method name as a key to selecting a handler, whereas mutating is more just metadata about the handler, and how it should be executed once it has been selected. Definitely a single handler couldn't be mutating and non-mutating for different methods, I agree.

> and the advantage though

To be clear, I'm not entirely sure there is one, other than removing some reflection which may not be an issue anyway :)

I guess my gut reaction is "oh, its annoying for new handlers to have to create two files when they might not need it", but given that new handlers are added pretty rarely, its a bit of a storm in a tea-cup I admit.

Reply:
@davidwengier now that I've made all the 'stateless' handlers just into one file, does that mitigate the issue?

Reply:
Yeah, I like that. The abstract class makes it even nicer :)